### PR TITLE
Skip unsupported digest algorithm

### PR DESCRIPTION
**Pull request by sauwming**
Report and patch by Alexander Traud:
" (some minor edits without altering its content)
See <https://stackoverflow.com/q/60683714>, found independently.

PJSIP does not support SHA-256 but only MD5 (and AKAv1-MD5; pjsip/sip_auth_client.c:respond_digest). However, the Backus-Naur-Form in RFC 3261 allows several WWW-Authenticate lines and the implementation has to pick its supported algorithm. RFC 3261 even known about unknown algorithms (token).

Currently, PJSIP 2.10, goes for the first WWW-Authenticate line only. If that is not supported, no progress (PJSIP_EINVALIDALGORITHM)

With RFC 7616, algorithms got added like SHA-256. Recently, the SIP service at sip.linphone.org added that. Give it a try and use my account "traud", password does not matter. You get two WWW-Authenticate lines, the first is SHA-256 and the second one is MD5.

I am *not* asking for supporting "SHA-256". I am asking for picking the right, your supported algorithm. This was required in RFC 3261 already. And actually my 12-year-old Nokia Mobile Phone picks MD5.

Note: pjsua command line tool faces the same issue.
"

This current patch has two limitations:
1. WWW-Authenticate without algorithm is ignored.
2. WWW-Authenticate with AKAv1-MD5 is ignored [1].
Both cases work(ed) further down in the stack in respond_digest(.).

### CODE REVIEW

**Comment by nanangizz** [comment]
`pjsip_auth_clt_reinit_req()` iterates all auth headers already, it just returns error immediately when finding an algo is not supported, so next auth header is not evaluated. So, I think we'd need to consult the standard first: when receiving multiple auth headers (with same realm?), should PJSIP respond to all algo (current behavior) or supported algo only?

**Comment by traud** [comment]
https://tools.ietf.org/html/rfc7616#section-3.9.1
example is explained in general in chapter 3.7. Furthermore, there is the definition of ‘algorithm’: 
> If the algorithm is not understood, the challenge SHOULD be ignored (and a different one used, if there is more than one).

This sentence is the same in the new RFC 7616 and the old RFC 2617.

**Comment by nanangizz** [comment]
Just proposed an alternate solution, please check :)

**Comment by traud** [comment]
No, that approach that does not work (correctly). Just tested, to double-check.

I used that approach (or at least a similar one) at start as well. That approach creates additional SIP requests. For example, pjsua invites someone and my server has MD5 as the second algorithm: pjsua creates not two but *four* INVITEs. Not the second but (only) the *fourth* INVITE contains a header `Authorization: Digest …`. Although that authenticates then successfully, such a behavior might have triggered suspicious filterers like [fail2ban](https://en.wikipedia.org/wiki/Fail2ban).

To be honest, I do not know what exactly those two additional SIP requests create. Therefore, I cannot say what is wrong with your approach. At start, just reading the code, I went for the same approach like you. Did you test it against _sip.linphone.org_? What is wrong with my approach? I can enhance it for the two cases no-algorithm and AKAv1-MD5. I did not do it because I am not that steady in PJStr comparison (and I am not sure where to move that AKAv1-MD5 constant then).

**Comment by nanangizz** [comment]
Just tested with _sip.linphone.org_, but unfortunately it responded (for both `REGISTER` & `INVITE` requests) with one authenticate header line only (?). The patch work fine with just added SIPp test scripts though, no multiple `REGISTER`.

Your patch is generally fine (minor issue, missing scheme check), but it adds more place doing scheme/supported-algo checks, currently it is done in `process_auth()` (and also further down in `auth_respond()` & `respond_digest()`). Basically, just feel better to leave the algo check in `process_auth()` as is now.

**Comment by traud** [comment]
Go for another, a third
`WWW-Authenticate: Digest … algorithm=MD2 …`
Yes, this MD-Two is a fictive but possible algorithm with the same name length as MD-Five.

**Comment by nanangizz** [comment]
Good test, found a bug with the MD2! The Authorization header generated was okay (REGISTER retried with only one Auth header, etc), but the error message was `Authorization failed ... server rejected with stale=false` while it should be `Unsupported digest algorithm "MD2"`.

**Comment by traud** [comment]
Works and passes my tests. Magically (?) those additional SIP requests are gone as well.

Nevertheless, I do not understand that whole code file. For example:
1. `process_auth(.)` returns many more status than just `PJSIP_EINVALIDALGORITHM`. But *all* of them seem to be silenced now after your patch.
2. What happens if the *realms* are different? If I am not totally misreading the code for `cached_auth =`, the realm of the first challenge is used always.

I have not tested this, yet. I understood my patch. It made sense. Now with the current patch, I do not understand anything. I cannot judge on yours, except that it works now for my test cases. Sorry. Cannot help here further.

**Comment by nanangizz** [comment]
> Works and passes my tests. Magically (?) those additional SIP requests are gone as well.

Great! Not sure what might cause the additional SIP requests, the code seems to generate only one request to me.

> Nevertheless, I do not understand that whole code file. For example:
> 
> 1. `process_auth(.)` returns many more status than just `PJSIP_EINVALIDALGORITHM`. But _all_ of them seem to be silenced now after your patch.

Yes, but as the function is actually still invoked for each auth challenge, any log will still be printed. And as the point seems to generate a valid auth respond, ignoring error of other (unsupported) challenge should be fine, and when no algo is supported, it will return `PJSIP_EAUTHNOAUTH` instead of `PJSIP_EAUTHNOCHAL` (with the original patch). Note that with the original patch, the function will be completely skipped for any unsupported algo.

> 2. What happens if the _realms_ are different? If I am not totally misreading the code for `cached_auth =`, the realm of the first challenge is used always.

Not sure which part of the code points to the first challenge.

> I have not tested this, yet. I understood my patch. It made sense. Now with the current patch, I do not understand anything. I cannot judge on yours, except that it works now for my test cases. Sorry. Cannot help here further.

Basically the idea is the same, it's just that the new patch is letting the algo check (among other checks) to be done inside `process_auth()`.

Thank you for the report, the patch, and all the helps.

**Comment by traud** [comment]
Not sure if you learned whether such discussion|sentence style is successful and therefore you do it. Are you interested why exactly the four SIP requests were created? I am not sure if that statement of you was curiosity = question; or simply a do not care anymore.

> Which part of the code points to the first challenge?

`cached_auth = … hchal->challenge.common.realm …`
I hope at least that was a question. I tested your change, and it does not use the first realm over the wire. However, I do not understand what happens there in that code part. `cached_auth` is initialized with the first, possible wrong challenge. Where is that overridden? That was the reason I did my change before `cached_auth` gets assigned.

> log […] still be printed

No question either. Are you interested? Now, I have to _defend_ my curiosity. Wrong, isn’t it?

I am not about the log messages but the return/status value. Previously, the status of `process_auth(.)` was forwarded, for example, `pjsip_auth_clt_reinit_req(.)` did return `PJSIP_EINVALIDALGORITHM`, which still can happen. Now, you changed the API to return just 
a) `PJ_SUCCESS`
b) `PJSIP_EAUTHNOCHAL`
c) `PJSIP_EAUTHNOAUTH`
The latter is even a new status which was only used in `pjsip_auth_srv_verify(.)` before.

This could affect upstream projects, for example, Digium Asterisk creates [own log messages …](https://github.com/asterisk/asterisk/blob/master/res/res_pjsip_outbound_authenticator_digest.c#L153)

**Comment by nanangizz** [comment]
> Not sure if you learned whether such discussion|sentence style is successful and therefore you do it. Are you interested why exactly the four SIP requests were created? I am not sure if that statement of you was curiosity = question; or simply a do not care anymore.
> 

Actually with your last sentence in [this comment](https://github.com/pjsip/pjproject/pull/2408#issuecomment-622012362), I thought it was you that didn't want to continue the discussion (for whatever reason), hence I said thanks and did such sentence style :)

The four SIP request was not reproduced here, so as mentioned before, I checked the code, but could not see how four SIP requests may be generated. If you provide a log to investigate or even have a ready explanation, I am definitely open and all ears, simple, no need to say things with such tone.

I decided to propose an alternative patch because problems in the original patch:
- more places doing supported algo check. If SHA-256 is implemented in the future, with original patch, both challenges MD5 & SHA-265 will go through `cached_auth=...`, `process_auth()` and so on, the same way as the alternative patch, so why not pass all challenges and let `process_auth()` to filter for things such as supported algo (as it is also already doing it).
- minor: limitation you mentioned (AKA-MD5)
- minor: directly accessing digest without scheme check

Please note that I don't mind at all to go back with the approach in the original patch if it is indeed the correct one.

> > Which part of the code points to the first challenge?
> 
> `cached_auth = … hchal->challenge.common.realm …`
> I hope at least that was a question. I tested your change, and it does not use the first realm over the wire. However, I do not understand what happens there in that code part. `cached_auth` is initialized with the first, possible wrong challenge. Where is that overridden? That was the reason I did my change before `cached_auth` gets assigned.
> 

AFAIK, `cached_auth` is supposed to be unique by realm and it initializes/updates its challenge info (e.g: nonce, nc) when that challenge is accepted (authorization header is generated for that challenge). Should you see any problem with patch doing on `cached_auth`, feel free to describe.

I think `pjsip_auth_clt_reinit_req()` should only accept one challenge for each realm, and this is still missing from the code (and any patch). Any comment on this?

> > log […] still be printed
> 
> No question either. Are you interested? Now, I have to _defend_ my curiosity. Wrong, isn’t it?
> 
> I am not about the log messages but the return/status value. Previously, the status of `process_auth(.)` was forwarded, for example, `pjsip_auth_clt_reinit_req(.)` did return `PJSIP_EINVALIDALGORITHM`, which still can happen. Now, you changed the API to return just
> a) `PJ_SUCCESS`
> b) `PJSIP_EAUTHNOCHAL`
> c) `PJSIP_EAUTHNOAUTH`
> The latter is even a new status which was only used in `pjsip_auth_srv_verify(.)` before.
> 
> This could affect upstream projects, for example, Digium Asterisk creates [own log messages …](https://github.com/asterisk/asterisk/blob/master/res/res_pjsip_outbound_authenticator_digest.c#L153)

Ok, got your point here. If no auth hdr is generated/updated, returning any last error should do it I guess.

**Comment by sauwming** [comment]
The latest patch was confirmed to work in an independent test by George Joseph (Asterisk).

**Review by nanangizz — APPROVED** [review]

**Review by trengginas — APPROVED** [review]
